project3.cc

```diff
@@ -0,0 +1,599 @@
+    newnode->_rchild = makeThread(_header);
+    newnode->_value = value;
+    newnode->_key = key;
+  }
+  else
+  {
+    ThreadedBinarySearchTree::insertr(key, value, _header->_lchild);
+  }
+}
+
+
+/*
+ * Recursive insertion called by insert()
+ * PARAM-key: the string used to determine tree position
+ * PARAM-value: value that for the node
+ * PARAM-root: the starting node for each recursion
+ */
+void ThreadedBinarySearchTree::insertr(string key, int value, Node *& root)
+{
+  if (isThread(root->_lchild) && (key < root->_key))
+  {
+    Node *newnode = new Node();
+
+    newnode->_lchild = root->_lchild;
+    newnode->_rchild = makeThread(root);
+
+    root->_lchild = newnode;
+
+    newnode->_value = value;
+    newnode->_key = key;
+  }
+  else if (isThread(root->_rchild) && (key > root->_key))
+  {
+    Node *newnode = new Node();
+
+    newnode->_lchild = makeThread(root);
+    newnode->_rchild = root->_rchild;
+
+    root->_rchild = newnode;
+
+    newnode->_value = value;
+    newnode->_key = key;
+  }
+  else
+  {
+    if (key > root->_key)
+    {
+      insertr(key, value, root->_rchild);
+    }
+
+    else if (key < root->_key)
+    {
+      insertr(key, value, root->_lchild);
+    }
+  }
+}
+
+
+void ThreadedBinarySearchTree::erase(Iterator iter)
+{
+  // if (empty())
+  // {
+  //         return end();
+  // }
+  // else
+  // {
+  //         Node *tempnode;
+  //         Node *parent = this->parent()._ptr;
+  //         Iterator it;
+  //         for (iter = this->inorder_begin(); iter != this->end(); it++)
+  //         {
+  //                 if (key == iter._ptr->_key)
+  //                 {
+  //                         tempnode = iter._ptr;
+  //
+  //                         if(!isThread(tempnode->_rchild) && isThread(tempnode->_lchild))
+  //                                 delete iter._ptr;
+  //                         else if (isThread(tempnode->_rchild) && !isThread(tempnode->_lchild))
+  //                                 delete iter._ptr;
+  //                         else if(isThread(tempnode->_rchild) && isThread(tempnode->_lchild))
+  //                                 delete iter._ptr;
+  //                         else
+  //                                 delete iter._ptr;
+  //
+  //                 }
+  //         }
+  //         return end();
+  // }
+  return end();
+}
+
+
+/*
+* Determine if the tree is empty
+* RETURN-true: if there is only a header
+* RETRUN-false: if there is atleast one node
+*/
+bool ThreadedBinarySearchTree::empty() const
+{
+  if (makePointer(_header->_lchild) == _header)
+  {
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*
+* Lookup function called the user clicks l
+* PARAM-key: the string used to determine tree position
+*/
+Iterator ThreadedBinarySearchTree::lookup(string key) const
+{
+  if (empty())
+  {
+    return end();
+  }
+  else
+  {
+    Iterator it;
+    for (it = this->inorder_begin(); it != this->end(); it++)
```

Comment:
Why do a linear search, O(n), here, when you do a binary search, O(log n) if the tree is reasonably balanced, in insert()?  This is not efficient.